### PR TITLE
Handle the case when mongodb returns less documents than cursor reque…

### PR DESCRIPTION
**Pull request by VictorDenisov**
…sted

If we request certain amount of values from a cursor the mongo db server can
return less than requested. So, if we provide -100000 then mongodb may
return 97899 and close the cursor.

Instead of relying on negative values this implementation will use only
positive numbers and will close the cursor itself as soon as the driver
receives enough results.

It fixes the issue #24 from github.

### CODE REVIEW

**Comment by gregwebs** [comment]
Thanks for fixing. Looks like the Travis build is failing right now

**Comment by VictorDenisov** [comment]
Hm. Looking into it.

**Comment by VictorDenisov** [comment]
The issues are fixed. Added the test that failed initially.

**Comment by VictorDenisov** [comment]
Can someone review and merge it?

File: Database/MongoDB/Query.hs [hunk 1]
```diff
@@ -582,20 +583,25 @@ nextBatch :: (MonadIO m, MonadBaseControl IO m) => Cursor -> Action m [Document]
...
     -- Pre-fetch next batch promise from server and return current batch.
-    Batch limit cid docs <- fulfill' fcol batchSize dBatch
-    dBatch' <- if cid /= 0 then nextBatch' fcol batchSize limit cid else return $ return (Batch 0 0 [])
-    return (dBatch', docs)
+    Batch mLimit cid docs <- fulfill' fcol batchSize dBatch
+    let newLimit = do
+              limit <- mLimit
+              return $ limit - (min limit $ fromIntegral $ length docs)
+    dBatch' <- if cid /= 0 && newLimit /= (Just 0) then nextBatch' fcol batchSize newLimit cid else return $ return (Batch (Just 0) 0 [])
+    when (newLimit == (Just 0)) $ unless (cid == 0) $ send [KillCursors [cid]]
```
Comment (gregwebs) [inline Database/MongoDB/Query.hs]: should combine `unless` with `when`

File: Database/MongoDB/Query.hs [hunk 2]
```diff
@@ -582,20 +583,25 @@ nextBatch :: (MonadIO m, MonadBaseControl IO m) => Cursor -> Action m [Document]
...
+    return (dBatch', maybe docs (\l -> take (fromIntegral l) docs) mLimit)
```
Comment (gregwebs) [inline Database/MongoDB/Query.hs]: I would probably write this in this style: `(maybe id (take . fromIntegral) mLimit) docs`

Comment (gregwebs) [inline Database/MongoDB/Query.hs]: Actually, I would probably write the whole thing as 

+    let newLimit = do
+              limit <- mLimit
+              return $ limit - (min limit $ fromIntegral $ length docs)
+    dBatch' <- if cid /= 0 && newLimit /= (Just 0) then nextBatch' fcol batchSize newLimit cid else return $ return (Batch (Just 0) 0 [])
+    when (newLimit == (Just 0)) $ unless (cid == 0) $ send [KillCursors [cid]]

Comment (gregwebs) [inline Database/MongoDB/Query.hs]: Actually, I would probably write the whole thing as something like this. Right now the logic is hard to follow

``` haskell
let emptyBatch = Batch (Just 0) 0 []
let getNextBtach = nextBatch' fcol batchSize newLimit cid
let takeDocs i = take $ fromIntegral i $ docs
case (cid, newLimit) of
    (0, Just 0) ->  (emptyBatch, takeDocs 0)
    (i, Just 0) ->  do
      send [KillCursors [cid]]
      (, takeDocs 0) <$> getNextBatch
    (_, Nothing) -> (, docs) <$> getNextBatch
```

File: Database/MongoDB/Query.hs [hunk 3]
```diff
@@ -524,29 +525,29 @@ queryRequest isExplain Query{..} = do
...
-        else (- fromIntegral limit, 1)
- where batchSize' = if batchSize == 1 then 2 else batchSize
-    -- batchSize 1 is broken because server converts 1 to -1 meaning limit 1
+batchSizeRemainingLimit batchSize mLimit =
+  case mLimit of
+    Nothing    -> (fromIntegral batchSize, Nothing)
+    Just limit ->
+      if 0 < batchSize && batchSize < limit
+        then (fromIntegral batchSize, Just limit)
+        else (fromIntegral limit, Just limit)
```
Comment (gregwebs) [inline Database/MongoDB/Query.hs]: here the `mLimit` is not changing. So

``` haskell
let remaining = case mLimit of
  Nothing -> batchSize
  Just limit -> if 0 < batchSize && batchSize < limit then batchSize else limit
in (fromIntegral remaining, mLimit)
```

**Comment by VictorDenisov** [comment]
Thanks Greg for the review. I'll update the request shortly.

File: Database/MongoDB/Query.hs (hunk 1 above)
Comment (VictorDenisov) [inline Database/MongoDB/Query.hs]: done

File: Database/MongoDB/Query.hs (hunk 2 above)
Comment (VictorDenisov) [inline Database/MongoDB/Query.hs]: done

File: Database/MongoDB/Query.hs (hunk 3 above)
Comment (VictorDenisov) [inline Database/MongoDB/Query.hs]: done

**Comment by gregwebs** [comment]
That's better!

**Comment by gregwebs** [comment]
want to put this on hackage? What is your user account there?

**Comment by VictorDenisov** [comment]
My account is VictorDenisov

**Comment by gregwebs** [comment]
You should have the ability to upload to hackage now

**Comment by VictorDenisov** [comment]
Thanks. I assume I need to bump the version to 2.0.6. I'm going to add a changelog file and then publish the package. Do you want me to make a pull request for this or can I just push the commits directly? Do you want to see the package candidate before I make it public?

**Comment by gregwebs** [comment]
I usually don't make Pull Requests for those things. I also don't have code to test out a package candidate with other than the persistent (persistent-mongodb) test suite. Although, that would actually be a great way to setup an automated test on TravisCI: take the setup from persistent, but remove all the other backends, and use mongoDB from source.

**Comment by VictorDenisov** [comment]
Could you please review this package candidate: http://hackage.haskell.org/package/mongoDB-2.0.6/candidate ?

**Comment by gregwebs** [comment]
this is my review: https://github.com/mongodb-haskell/mongodb/pull/27
